openscm/timeseries_converter.py

```diff
@@ -0,0 +1,364 @@
+    taken as the average of adjacent periods, values at the period middles are taken
+    such that the integral over a period is the same as for the input data.
+
+    Parameters
+    ----------
+    values
+        Timeseries values of period averages
+
+    Returns
+    -------
+    np.ndarray
+        Values of linearization (of length ``2 * len(values) + 1``)
+    """
+    edge_point_values = (values[1:] + values[:-1]) / 2
+    middle_point_values = (
+        4 * values[1:-1] - edge_point_values[:-1] - edge_point_values[1:]
+    ) / 2
+    # values = (
+    #   1 / 2 * (edges_lower + middle_point_values) * 1 /2
+    #   + 1 / 2 * (middle_point_values + edges_upper)
+    # ) / 2
+    first_edge_point_value = (
+        2 * values[0] - edge_point_values[0]
+    )  # values[0] = (first_edge_point_value + edge_point_values[0]) / 2
+    last_edge_point_value = (
+        2 * values[-1] - edge_point_values[-1]
+    )  # values[-1] = (last_edge_point_value + edge_point_values[-1]) / 2
+    return np.concatenate(
+        (
+            np.array(
+                [
+                    np.concatenate(([first_edge_point_value], edge_point_values)),
+                    np.concatenate(([values[0]], middle_point_values, [values[-1]])),
+                ]
+            ).T.reshape(2 * len(values)),
+            [last_edge_point_value],
+        )
+    )
+
+
+class TimeseriesConverter:
+    """
+    Converts timeseries and their points between two timeseriess (each defined by a time
+    of the first point and a period length).
+    """
+
+    _source: np.ndarray
+    """Source timeseries time points"""
+
+    _target: np.ndarray
+    """Target timeseries time points"""
+
+    _timeseries_type: ParameterType
+    """Time series type"""
+
+    _interpolation_type: InterpolationType
+    """Interpolation type"""
+
+    _extrapolation_type: ExtrapolationType
+    """Extrapolation type"""
+
+    def __init__(
+        self,
+        source_time_points: np.ndarray,
+        target_time_points: np.ndarray,
+        timeseries_type: ParameterType,
+        interpolation_type: InterpolationType,
+        extrapolation_type: ExtrapolationType,
+    ):
+        """
+        Initialize.
+
+        Parameters
+        ----------
+        source_time_points
+            Source timeseries time points
+        target_time_points
+            Target timeseries time points
+        timeseries_type
+            Time series type
+        interpolation_type
+            Interpolation type
+        extrapolation_type
+            Extrapolation type
+        """
+        self._source = np.array(source_time_points, copy=True)
+        self._target = np.array(target_time_points, copy=True)
+        self._timeseries_type = timeseries_type
+        self._interpolation_type = interpolation_type
+        self._extrapolation_type = extrapolation_type
+
+        if source_time_points[0] > target_time_points[1]:
```

Comment:
this will become conditional on extrapolation_type in future right?

Reply:
Yes, will add a comment